src/xml2_node.cpp

```diff
@@ -27,13 +27,57 @@ std::string nodeName(T* node, SEXP nsMap) {
   return prefix + ":" + name;
 }
 
+SEXP node_name_impl(SEXP x, SEXP nsMap) {
+  NodeType type = getNodeType(x);
+
+  SEXP out;
+
+  switch(type) {
+  case NodeType::missing:
+    out = NA_STRING;
+    break;
+  case NodeType::node: {
+    SEXP node_sxp = VECTOR_ELT(x, 0);
+    XPtrNode node(node_sxp);
+
+    std::string name = nodeName(node.checked_get(), nsMap);
+    out = Rf_mkCharLenCE(name.c_str(), name.size(), CE_UTF8);
```

Comment:
I suppose this has always been the case, but if we have an R error anywhere in here, then it will skip over any C++ destructors (like for XPtrNode, if it has one, and `std::string`) on the way out

I don't think there is a reason to try and improve this in this PR, but I will keep an eye out for potentially problematic cases as I continue my review

Reply:
Yeah, we have an odd mix of C and C++ in this package which we should try and fix at some point.